src/Lumper.Lib/BSP/BspFile.cs

```diff
@@ -35,45 +43,238 @@ public class BspFile
+        catch (Exception ex)
+        {
+            Logger.Error(ex, "Failed to write BSP file to stream");
+            success = false;
+        }
+
+        if (success)
+        {
+            if (overwritingOpenFile)
+            {
+                // Dispose of filestream we were just reading from
+                FileStream!.Dispose();
+
+                // Reopen, FileMode.Create overwrites
+                using FileStream fstream = File.Open(outPath, FileMode.Create);
+                stream.Seek(0, SeekOrigin.Begin);
+
+                // Copy memorystream we just wrote to over to new filestream, dispose
+                stream.CopyTo(fstream);
+                // Dispose of writer and underlying `stream`.
+                writer.Dispose();
+            }
+            else
+            {
+                // This is a filestream to a new file, just flush and close.
+                writer.Dispose();
+            }
+
+            // Update paths and name, and reopen the file we just wrote to, not reading anything.
+            // This means we keep an open stream to the file (stops other processes tampering), and
+            // can use it for unmanaged lump/pakfile (in some cases) when next saving, without
+            // actually loading anything into memory yet.
+            Name = Path.GetFileNameWithoutExtension(outPath);
+            FilePath = outPath;
+            FileStream = File.OpenRead(outPath);
+
+            Logger.Info($"Saved {outPath}");
+
+            // We just updated our filestream, and saving might've changed headers,
+            // so update the stream and headers for any lump which cares about a
+            // reference to the filestream (unmanaged lumps and the pakfile).
+            foreach ((BspLumpType bspLumpType, BspLumpHeader bspHeader) in writer.LumpHeaders)
+            {
+                Lump lump = GetLump(bspLumpType);
+
+                if (lump is GameLump gl)
+                {
+                    foreach ((Lump glLump, LumpHeaderInfo? glLumpHeader) in gl.Reader.Lumps)
+                    {
+                        if (glLump is not IFileBackedLump glfbLump)
+                            continue;
+
+                        glfbLump.DataStream = FileStream;
+                        glfbLump.DataStreamLength = glLumpHeader.Length;
+                        glfbLump.DataStreamOffset = glLumpHeader.Offset;
+                    }
+                }
+
+                if (lump is not IFileBackedLump fbLump)
+                    continue;
+
+                fbLump.DataStream = FileStream;
+                fbLump.DataStreamLength = bspHeader.Length;
+                fbLump.DataStreamOffset = bspHeader.Offset;
+            }
+
+            return;
+        }
+
+        // Save failed. Stream we were just writing to is useless, dispose.
+        // Save operation doesn't alter this BSP or underlying filestream if open.
+        writer.Dispose();
+
+        if (backupPath is null)
+            return;
```

Comment:
don't like multiple returns in a function .. especially when they are so far in (`if null return` or something like this in the beginning is fine)

Reply:
Fair enough. I think it's a point of preference and I'm more pro using `return`s for control flow than you, given the size of this function I suppose it makes sense to indent rather than early return.

Note though that we have CSharpier incoming with more aggressive line wrapping, and heavy indentation can really affect readability as well. In some case splitting the function into multiple is better, but I see a lot of JS code with horrible readability because they've broken the function up too much to avoid indentation - in those cases I think early returning tends to be much better.